internal/http/services/owncloud/ocs/handlers/apps/sharing/shares/shares.go

```diff
@@ -1549,84 +1535,128 @@ func (h *Handler) updatePublicShare(w http.ResponseWriter, r *http.Request, shar
+	if ok {
+		updatesFound = true
+		var newExpiration *types.Timestamp
+		if expireTimeString[0] != "" {
+			newExpiration, err = parseTimestamp(expireTimeString[0])
+			if err != nil {
+				response.WriteOCSError(w, r, response.MetaBadRequest.StatusCode, "invalid datetime format", err)
+				return
+			}
+		}
+
+		beforeExpiration, _ := json.Marshal(before.Share.Expiration)
+		afterExpiration, _ := json.Marshal(newExpiration)
+		if string(afterExpiration) != string(beforeExpiration) {
+			logger.Debug().Str("shares", "update").Msgf("updating expire date from %v to: %v", string(beforeExpiration), string(afterExpiration))
+			updates = append(updates, &link.UpdatePublicShareRequest_Update{
+				Type: link.UpdatePublicShareRequest_Update_TYPE_EXPIRATION,
+				Grant: &link.Grant{
+					Expiration: newExpiration,
+				},
+			})
+		}
 	}
 
 	// Password
-	if len(r.FormValue("password")) > 0 {
+	newPassword, ok := r.Form["password"]
+	// update or clear password
+	if ok {
+		updatesFound = true
 		logger.Info().Str("shares", "update").Msg("password updated")
 		updates = append(updates, &link.UpdatePublicShareRequest_Update{
 			Type: link.UpdatePublicShareRequest_Update_TYPE_PASSWORD,
 			Grant: &link.Grant{
-				Password: r.FormValue("password"),
+				Password: newPassword[0],
 			},
 		})
 	}
 
-	// Updates are atomical. See: https://github.com/cs3org/cs3apis/pull/67#issuecomment-617651428 so in order to get the latest updated version
-	if len(updates) == 0 {
-		response.WriteOCSError(w, r, response.MetaBadRequest.StatusCode, "There is nothing to update.", nil) // TODO(refs) error? a simple noop might suffice
-		return
-	}
+	publicShare := before.Share
 
-	req := &link.UpdatePublicShareResponse{}
-	for k := range updates {
-		req, err = gwC.UpdatePublicShare(r.Context(), &link.UpdatePublicShareRequest{
-			Ref: &link.PublicShareReference{
-				Spec: &link.PublicShareReference_Id{
-					Id: &link.PublicShareId{
-						OpaqueId: shareID,
+	// Updates are atomical. See: https://github.com/cs3org/cs3apis/pull/67#issuecomment-617651428 so in order to get the latest updated version
+	if len(updates) > 0 {
+		uRes := &link.UpdatePublicShareResponse{Share: before.Share}
+		for k := range updates {
+			uRes, err = gwC.UpdatePublicShare(r.Context(), &link.UpdatePublicShareRequest{
+				Ref: &link.PublicShareReference{
+					Spec: &link.PublicShareReference_Id{
+						Id: &link.PublicShareId{
+							OpaqueId: shareID,
+						},
 					},
 				},
-			},
-			Update: updates[k],
-		})
-		if err != nil {
-			log.Err(err).Str("shareID", shareID).Msg("sending update request to public link provider")
+				Update: updates[k],
+			})
+			if err != nil {
+				log.Err(err).Str("shareID", shareID).Msg("sending update request to public link provider")
+				response.WriteOCSError(w, r, response.MetaServerError.StatusCode, "Error sending update request to public link provider", err)
+				return
+			}
 		}
+		publicShare = uRes.Share
+	} else if !updatesFound {
```

Comment:
if I follow the logic correctly, isn't `updatesFound` unnecessary? the reverse of `len(updates) > 0` is essentially `len(updates) == 0`

Reply:
updatesFound is only there to trigger the 400 Bad Request in case none of the expected attributes has been passed in, basically "no update has been requested".

if you do specify a value, even if it matches already what's in the DB, then it's accepted as valid. before this fix it was saying that "setting permissions to 15 when it's already 15" is a bad request as it triggers no updates. faulty logic

Reply:
because when you set permissions to 15 when they're already 15, the updates array will be empty as we have difference checks above.

Reply:
I mean the existence of the `updatesFound` variable, not the execution path :p

Reply:
the semantics of `updatesFound` is not the same as `len(updates)`. you can have a case where zero updates exist in the array but still some updates have been provided through attributes

Reply:
👍 works as expected. Clarified the misunderstanding.